File: python/taflow/indicators/trade_volume_index.py
"""Public adapter for the native Trade Volume Index state."""

from typing import Any

import numpy as np

from .._native import TradeVolumeIndex as _Native
from .._series import as_float64_series


class TradeVolumeIndex:
    """Accumulate volume using a persistent minimum-tick price direction.

    A close change above ``min_tick`` selects positive volume and a change
    below ``-min_tick`` selects negative volume. Smaller changes retain the
    previous direction. The first bar seeds its close and outputs ``NaN``.
    This class maps to Wickra ``TradeVolumeIndex``.

    Args:
        close: Initial chronological closing prices.
        volume: Initial chronological volumes.
        min_tick: Non-negative direction threshold, default 0.25.

    Raises:
        ValueError: If inputs are misaligned or ``min_tick`` is invalid.
    """

    def __init__(self, close: Any, volume: Any, min_tick: float = 0.25) -> None:
        """Initialize the state and process the supplied close/volume history."""
        self._state = _Native(min_tick)
        self.extend(close, volume)

    def append(self, close: float, volume: float) -> "TradeVolumeIndex":
        """Append one close/volume sample and return this instance."""
        self._state.append(float(close), float(volume))
        return self

    def extend(self, close: Any, volume: Any) -> "TradeVolumeIndex":
        """Append aligned close and volume histories and return this instance."""
        close_series = as_float64_series(close)
        volume_series = as_float64_series(volume)
        if len(close_series) != len(volume_series):
            raise ValueError("close and volume must have equal lengths")
        self._state.extend(close_series, volume_series)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest cumulative index, or ``None`` after no output."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return aligned TVI history, including the initial warm-up ``NaN``."""
        return self._state.compute()

    def reset(self) -> "TradeVolumeIndex":
        """Reset close, direction, and total, then return this instance."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of samples stored by the native state."""
        return len(self._state)

File: crates/taflow-core/src/indicators/trade_volume_index.rs
use crate::error::{TaError, TaResult};

/// Cumulative volume signed by a persistent minimum-tick direction.
#[derive(Debug, Clone)]
pub struct TradeVolumeIndex {
    min_tick: f64,
    previous_close: Option<f64>,
    direction: f64,
    total: f64,
    value: Option<f64>,
}

impl TradeVolumeIndex {
    /// Create a TVI state with a finite, non-negative minimum tick.
    pub fn new(min_tick: f64) -> TaResult<Self> {
        if !min_tick.is_finite() || min_tick < 0.0 {
            return Err(TaError::InvalidParameter {
                name: "min_tick",
                value: min_tick.to_string(),
                reason: "must be finite and non-negative",
            });
        }
        Ok(Self {
            min_tick,
            previous_close: None,
            direction: 0.0,
            total: 0.0,
            value: None,
        })
    }

    /// Append one close/volume sample and return the cumulative TVI.
    pub fn append(&mut self, close: f64, volume: f64) -> Option<f64> {
        let Some(previous) = self.previous_close.replace(close) else {
            self.value = None;
            return None;
        };
        let change = close - previous;
        if change > self.min_tick {
            self.direction = 1.0;
        } else if change < -self.min_tick {
            self.direction = -1.0;
        }
        self.total += self.direction * volume;
        self.value = Some(self.total);
        self.value
    }

    /// Return the latest TVI, or `None` before the second sample.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Clear the close reference, direction, cumulative total, and latest value.
    pub fn reset(&mut self) {
        self.previous_close = None;
        self.direction = 0.0;
        self.total = 0.0;
        self.value = None;
    }
}
